src/Multicall3.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.12;
+
+/// @title Multicall3
+/// @notice Aggregate results from multiple read-only function calls
+/// @dev Multicall & Multicall2 backwards-compatible
+/// @author Michael Elliot <mike@makerdao.com>
+/// @author Joshua Levine <joshua@makerdao.com>
+/// @author Nick Johnson <arachnid@notdot.net>
+/// @author Andreas Bigger <andreas@nascent.xyz>
+
+contract Multicall3 {
+    struct Call {
+        address target;
+        bytes callData;
+    }
+
+    struct Call3 {
+        address target;
+        bool allowFailure;
+        bytes callData;
+    }
+
+    struct Result {
+        bool success;
+        bytes returnData;
+    }
+
+    function aggregate(Call[] calldata calls) public returns (uint256 blockNumber, bytes[] memory returnData) {
+        blockNumber = block.number;
+        uint256 length = calls.length;
+        returnData = new bytes[](length);
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(success, "Multicall3: aggregate failed");
+            returnData[i] = ret;
+            unchecked { ++i; }
+        }
+    }
+
+    function tryAggregate(bool requireSuccess, Call[] calldata calls) public returns (Result[] memory returnData) {
+        uint256 length = calls.length;
+        returnData = new Result[](length);
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            if (requireSuccess) require(success, "Multicall3: tryAggregate failed");
+            returnData[i] = Result(success, ret);
+            unchecked { ++i; }
+        }
+    }
+
+    function tryBlockAndAggregate(bool requireSuccess, Call[] calldata calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = tryAggregate(requireSuccess, calls);
+    }
+
+    function blockAndAggregate(Call[] calldata calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        (blockNumber, blockHash, returnData) = tryBlockAndAggregate(true, calls);
+    }
+
+    function aggregate3(Call3[] calldata calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = new Result[](calls.length);
+        uint256 length = calls.length;
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(calls[i].allowFailure || success, "Multicall3: aggregate3 failed");
```

Comment:
@transmissions11 mentioned we can save gas in this check by instead using assembly's `or()` operator to avoid short-circuiting. This would remove a JUMPI which costs 10 gas, so just would need to verify that 10 gas is more than the cost to evaluate the right-hand side of the expression

Reply:
Should this also use a custom error instead of an error string?

Reply:
> This would remove a JUMPI which costs 10 gas, so just would need to verify that 10 gas is more than the cost to evaluate the right-hand side of the expression

This is very likely the case.

Reply:
> Should this also use a custom error instead of an error string?

@ind-igo If it were my preference, we would, but using <=32 byte strings for backwards-compatibility. We're trying to optimize wrt to the tradeoff of backwards-compatibility <> optimizing.

Reply:
Would it also make sense to have `success` first in the evaluation, assuming that the user knows what they do so it has a higher chance to be correct and stop EVM from evaluating the mapping

Reply:
Hmm, yea I wonder which condition would be true more often. I'm inclined to think `success` meaning it should be first. But either way it sounds like using `or()` to remove a JUMPI will be cheapest, I just have not actually implemented and benchmarked it yet (maybe @abigger87 has?)

Reply:
> Hmm, yea I wonder which condition would be true more often. I'm inclined to think `success` meaning it should be first. But either way it sounds like using `or()` to remove a JUMPI will be cheapest, I just have not actually implemented and benchmarked it yet (maybe @abigger87 has?)

@mds1 I read about short-circuiting, but i can't grep why the assembly or will save a JUMPI. ELI5 pls ser 🥲

Reply:
@odyslam A solidity high level `if (a && b) {...}` gets converted to the following in Yul. 

```
if a {
    if b {
        ....
    }
}
```

This allows for short-circuiting. Assume that the expressions `a` and `b` are state changing, if `a` returns false, then `b` would never be executed. But if the expressions `a` and `b` are side effect free, then `or(a, b)` may be cheaper.
```
if or(a, b) {
    ....
}
```

Reply:
done in d2320fb